tool/public/client.js

```diff
@@ -9,28 +9,45 @@ class ConverterUi {
 
 		iptFile.addEventListener("change", () => {
 			const reader = new FileReader();
-			reader.onload = () => iptText.value = reader.result;
 			reader.readAsText(iptFile.files[0]);
-			doConvert();
+			reader.onload = () => {
+				let text;
+
+				// Check whether filetype is legal
+				if (iptFile.value.match(/\.(json|txt)$/i) || iptFile.value.match(/$(?<!\.\w+)/)) { // .json, .txt, or no filetype
+					text = reader.result;
+				} else if (iptFile.value.match(/\.db$/i)) { // .db
+					text = `[\n\t${reader.result.replace(/\}\n\{/g, "},\n\t{")}]`; // Hastily converting to JSON
+				} else {
+					outText.value = "Failed to parse input text!\n\n> Invalid filetype";
+					console.error(`Failed to load invalid filetype: ${iptFile.value.match(/(?<=\/|\\)[^\/\\]+$/)?.[0]}`); // This monstrosity just grabs the filename from the path
+					return;
+				}
+
+				iptText.value = text;
+				doConvert();
+			};
 		});
 
 		const doConvert = () => {
-			let json;
 			try {
-				json = JSON.parse(iptText.value);
+				outText.value = JSON.stringify(Converter.getConverted(JSON.parse(iptText.value)), null, "\t");
 			} catch (e) {
-				console.log("Failed to parse input text!");
+				outText.value = `Failed to parse input text!\n\n${e}`;
 				throw e;
 			}
-			outText.value = JSON.stringify(Converter.getConverted(json), null, "\t");
 		};
 
 		btnConvert.addEventListener("click", () => doConvert());
 
 		btnCopy.addEventListener("click", async () => {
 			await navigator.clipboard.writeText(outText.value);
+			btnCopy.innerHTML = "Copy ✓";
```

Comment:
does this change the size of the button? :^) if so, stick a fixed width on the button so the UI doesn't shuffle about

my 5et convention is to have reusable stuff for "common" widths, e.g. `w-100p { 100px; }`, so ideally stick one of those in the `<style>` to make later refactoring easier

Reply:
and if the exact choice of `Copy ✓` as a string is to make the button _not_ shift about, then this is a shitty hack/subtle hole for the next person to fall down, and it should be `Copied ✓` instead, with the above changes made

Reply:
Nah, the button does expand while showing the tick. I did consider this, but I neglected to do anything fancier because I expected this to be a niche thing for developers who won't be hammering the Copy button over and over again... and also since there's (currently) no buttons to the right of Copy it doesn't really matter 🤣

But you are right, so I'll stick in the fixed-width classes. Also, 'Copied' over 'Copy' is indeed a better choice; I just didn't think of it.